matview/dbutils.py

```diff
@@ -0,0 +1,125 @@
+from django.db import migrations
+
+from .models import MatView
+
+
+def _make_matview_migration(src, filters, name):
+    """Generate the SQL strings forward and backwards to create and drop materialized views.
+
+    This create a second view which simply materializes a count of the regular one. Both have
+    a unique index to allow concurrent updates.
+    """
+
+    data_clause = ("'{" + ','.join((
+        '"%s":"%s"' % (k, filters[k])
+        for k in sorted(filters)
+    )) + "}'")
+    count_only = not filters
+    query = 'select %s from %s where data @>%s' % ('*', src, data_clause)
+    main_matview_tmpl = """
+        CREATE MATERIALIZED VIEW %(name)s AS %(query)s;
+        CREATE UNIQUE INDEX %(name)s_pk ON %(name)s(id);
+    """
+    count_matview_tmpl = """
+        CREATE MATERIALIZED VIEW %(name)s__count AS SELECT 1 AS id, COUNT(*) FROM %(count_src)s;
+        CREATE UNIQUE INDEX %(name)s__count_pk ON %(name)s__count(id);
+    """
+
+    forward = ""
+    if not count_only:
+        forward = forward + (main_matview_tmpl % locals())
+        count_src = name
+    else:
+        count_src = src
+    forward = forward + (count_matview_tmpl % locals())
+
+    return migrations.RunSQL(
+        forward,
+        """
+        DROP MATERIALIZED VIEW IF EXISTS %(name)s__count CASCADE;
+        DROP MATERIALIZED VIEW IF EXISTS %(name)s CASCADE;
+        """ % locals()
+    )
+
+
+def get_matview_name(model, filters):
+    """Generate the name of a materialized view for a model and a given set of filters."""
+
+    app_label, model_name = model.split('.')
+    name = '_xx_'.join(
+        '%s_%s' % (k, filters[k])
+        for k in sorted(filters.keys())
+    )
+    name = 'matview_mv_%s_%s_%s' % (app_label, model_name, name)
+    return name.lower().strip('_')
```

Comment:
we can't support that in this form because we need the _ as a separator, so it can't be itself significant

Reply:
It sounds like you're saying that this would all break if someone used a filter like ``{'foo': '_bar_'}``.  If that's the case, we should either document it and fail if unacceptable parameters are passed in, or better, make it work regardless.